### PR TITLE
FEAT-#2800: add wait_computations, wait apis

### LINKED ISSUE #2800: Add wait function for _modin_frame
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. What kind of performance improvements would you like to see with this new API?

This is the most lightweight way to wait for the end of parallel computations on partitions.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2800 
- [x] tests added and passing

### CODE REVIEW

File: modin/engines/base/frame/partition_manager.py
Comment (anmyachev) [inline modin/engines/base/frame/partition_manager.py:563]: Maybe `flatten`?

**Review by Garra1980 — DISMISSED** [review]

File: modin/engines/base/frame/data.py
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:2105]: What would you think if we made this a decorator instead? Then decorate all of the methods that trigger computation. 

If it is implemented this way, we could add a config for `modin.config.BenmarkMode` and check within the decorator if that flag is set. What do you think?

**Review by devin-petersohn — COMMENTED** [review]
Looks great! One question on your thoughts for structure.

File: modin/engines/base/frame/data.py
Comment (anmyachev) [reply modin/engines/base/frame/data.py:2109]: I suggest leaving both the decorator and the function, since in different situations it is convenient to use either one or the other.

Comment (anmyachev) [reply modin/engines/base/frame/data.py:2109]: It seems that we should cast map object to list

File: asv_bench/benchmarks/utils.py
Comment (anmyachev) [inline asv_bench/benchmarks/utils.py:233]: I forgot that `map` changed behavior in python3.

Python3 - type(map return value) like iterator
Python2 - type(map return value) - list

**Comment by anmyachev** [comment]
@devin-petersohn ready for review

File: modin/engines/base/frame/data.py
Comment (devin-petersohn) [reply modin/engines/base/frame/data.py:2109]: I don't think we need both. In fact, I prefer this functionality not be exposed here. We can move the decorator into the frame manager.

I prefer these be deleted here because we don't want others to be accidentally calling them. Benchmark Mode is for internal and developer use only.

File: modin/engines/base/frame/partition_manager.py
Comment (devin-petersohn) [inline modin/engines/base/frame/partition_manager.py:559]: Along with changes above, I would move this outside the class as a decorator and decorate all functions, first checking if `BenchmarkMode.get()` then executing the line below. This way, it's not an API that is explicitly available for users.

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
I think there's a less invasive design explained below. Please let me know if it makes sense!

File: modin/engines/base/frame/partition_manager.py
Comment (anmyachev) [reply modin/engines/base/frame/partition_manager.py]: @devin-petersohn thanks for review! It's make sense. 

`wait_computations` was made as decorator function and other wait ways were removed. Please review again :)

**Review by devin-petersohn — COMMENTED** [review]
The approach looks great! I tested locally but I can't seem to get the synchronous mode to work. Here is how I tested:

```python
import modin.pandas as pd
import numpy as np
from modin.config import BenchmarkMode

BenchmarkMode.put(True)
df = pd.DataFrame(np.random.randint(0, 100, size=(2**20, 2**8))).add_prefix("col")
a = df.applymap(lambda x: x)  # returns immediately
```

Do I need to do something else?

**Comment by anmyachev** [comment]
> The approach looks great! I tested locally but I can't seem to get the synchronous mode to work. Here is how I tested:
> 
> ```python
> import modin.pandas as pd
> import numpy as np
> from modin.config import BenchmarkMode
> 
> BenchmarkMode.put(True)
> df = pd.DataFrame(np.random.randint(0, 100, size=(2**20, 2**8))).add_prefix("col")
> a = df.applymap(lambda x: x)  # returns immediately
> ```
> 
> Do I need to do something else?

No, it's enough.

I tested your script with `time` calls. I got next times: `3.08` sec in sync mode and 0.19 sec in async mode.
Environment: `MODIN_CPUS=44, MODIN_ENGINE=ray`

```python
import modin.pandas as pd
import numpy as np
from modin.config import BenchmarkMode
from time import time

BenchmarkMode.put(True)
df = pd.DataFrame(np.random.randint(0, 100, size=(2**20, 2**8))).add_prefix("col")
print(f"dataframe created")
start = time()
a = df.applymap(lambda x: x)
print(f"applymap: {time()-start}")

```

**Comment by anmyachev** [comment]
@devin-petersohn are you using the latest commit from the PR? What about modin environment variables? Which ones have you tested with?

**Comment by devin-petersohn** [comment]
@anmyachev This time I tried it, it is working. I do not know why it failed the first time, I was using the ProgressBar with it. I have noticed that the ProgressBar does not work properly if both are set to `True` because `BenchmarkMode` will block until the end and `ProgressBar` also blocks. Can you add a check for when `BenchmarkMode` and `ProgressBar` are enabled that throws an Exception for being incompatible?

**Comment by anmyachev** [comment]
> @anmyachev This time I tried it, it is working. I do not know why it failed the first time, I was using the ProgressBar with it. I have noticed that the ProgressBar does not work properly if both are set to `True` because `BenchmarkMode` will block until the end and `ProgressBar` also blocks. Can you add a check for when `BenchmarkMode` and `ProgressBar` are enabled that throws an Exception for being incompatible?

@devin-petersohn thanks for review! The check was be added.

**Review by devin-petersohn — APPROVED** [review]
Thanks @anmyachev LGTM

**Comment by devin-petersohn** [comment]
@modin-project/modin-omnisci waiting for your approval.

**Review by Garra1980 — APPROVED** [review]

File: modin/config/envvars.py
Comment (YarShev) [inline modin/config/envvars.py:272]: syncronous -> synchronous

File: modin/engines/base/frame/partition_manager.py
Comment (YarShev) [inline modin/engines/base/frame/partition_manager.py:45]: Why is this check necessary?

File: modin/experimental/engines/omnisci_on_ray/frame/partition.py
Comment (YarShev) [inline modin/experimental/engines/omnisci_on_ray/frame/partition.py:47]: Why don't we handle ray operations here as we do for pandas on ray? (see [here](https://github.com/modin-project/modin/pull/2801/files#diff-2df6c73d4a63d1c1345522920436cf6b0aec9abd1105fee23ce74a524981a35aR90))

File: modin/engines/base/frame/partition_manager.py
Comment (anmyachev) [reply modin/engines/base/frame/partition_manager.py:45]: To distinguish situations where one or more values are returned.

File: modin/experimental/engines/omnisci_on_ray/frame/partition.py
Comment (anmyachev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition.py:47]: I did in the same style as `get` function in `OmnisciOnRayFramePartition` class.

File: modin/engines/base/frame/partition_manager.py
Comment (YarShev) [reply modin/engines/base/frame/partition_manager.py:45]: Which situations do you mean? Please, give an example.

File: modin/experimental/engines/omnisci_on_ray/frame/partition.py
Comment (YarShev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition.py:47]: We probably need to do the same handling for `OmnisciOnRayFramePartition`'s operations that include ray calls.

File: modin/engines/base/frame/partition_manager.py
Comment (anmyachev) [reply modin/engines/base/frame/partition_manager.py:45]: In instance, `from_pandas` and `map_partitions`

Comment (YarShev) [reply modin/engines/base/frame/partition_manager.py:45]: I see, it makes sense for `from_pandas` but not for `map_partitions`. The latter returns np.array.

File: modin/experimental/engines/omnisci_on_ray/frame/partition.py
Comment (anmyachev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition.py:47]: https://github.com/modin-project/modin/issues/2837

**Pull request merged**
